### PR TITLE
feat(gatsby): Rewrite resource loading - retry mechanism

### PR DESCRIPTION
**Pull request by Moocar**
I won't be able to work on this for another week or so. Posting here in case anyone wants to take a stab at finishing it. Also a good opportunity to get feedback on the actual solution.

## Description

Rewrite of `loader.js`. Goals are:

- Handle case where the `XmlHttpRequest.readyState === 4`, and `statusCode === 0`
- If network error, or 500 error, or statusCode === 0, then retry 3 times before giving up
- Clean up code to remove ad hoc "if in development do this, otherwise do this". Solved with polymorphism. Not perfect, but hopefully easier to extend.
- Make loader instantiable. Reduce ad-hoc global state.
- Remove separate 404 functions. e.g `loadSyncOr404`. 
- enable future change where cached page-data can be expired. We could for instance add a "createdAt" field to entries in the `loader.pageDb`, and periodically expire them, or lazily re-request.

## Still TODO

- get tests working for `production-runtime` and `develop-runtime` end to end tests
- This is critical core functionality, and it has been rewritten, so MOAR TESTING!
- More comments/documentation

## Things I'm not happy with

- Circular dependency between api-runner and loader. So we have to contruct both, then set the api-runner onto the loader. This is how it's always been, but it would be nice to fix. Unfortunately, to do so, we'd have to make the api-runner constructable, which would require modifications to a large part of the codebase
- loader itself is still a global. Even though I've made it instantiable, we still need a global exported var since passing it everywhere would require change to too much of the codebase. Maybe one day
- I've never been a fan of `EnsureResources`, specifically how it uses the React lifecycle functions to cause loader retries. I tried removing it, but found out it's necessary to support forwards/back functionality after a refresh. It would be great to trigger resource loading directly off window.history events, possibly even removing Reach Router entirely. Maybe one day :)

## Related Issues

- Hopefully addresses some of https://github.com/gatsbyjs/gatsby/issues/11524

### CODE REVIEW

File: packages/gatsby/cache-dir/navigation.js [hunk 1]
```diff
@@ -82,6 +82,16 @@ const navigate = (to, options = {}) => {
...
 
   loader.loadPage(pathname).then(pageResources => {
+    // If no page resources, then refresh the page
+    // Do this, rather than simply `window.location.reload()`, so that
+    // pressing the back/forward buttons work - otherwise when pressing
+    // back, the browser will just change the URL and expect JS to handle
+    // the change, which won't always work since it might not be a Gatsby
+    // page.
+    if (!pageResources) {
+      window.history.replaceState({}, ``, location.href)
```
[the remark this answers is not in the document]
Comment (Moocar) [reply packages/gatsby/cache-dir/navigation.js]: I picked it up from the most excellent project Gatsbyjs! In [ensure-resources.js](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/cache-dir/ensure-resources.js#L67) on master ;)

File: packages/gatsby/cache-dir/loader.js [hunk 2]
```diff
@@ -105,273 +28,282 @@ const doFetch = (url, method = `GET`) =>
...
+        // This was just prefetched, so will return a response from
+        // the cache instead of making another request to the server
+        this.loadPageDataJson(pagePath)
+      )
+      .then(result => {
+        if (result.status !== `success`) {
+          return Promise.resolve()
+        }
+        const pageData = result.payload
+        // Tell plugins the path has been successfully prefetched
```
[the remark this answers is not in the document]
Comment (Moocar) [reply packages/gatsby/cache-dir/loader.js]: This comment was already in [loader.js](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/cache-dir/loader.js#L229) on master. I honestly don't fully understand it.

**Comment by Moocar** [comment]
Thanks for continuing the work on this @wardpeet! Much appreciated.

File: packages/gatsby/cache-dir/loader.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (KyleAMathews) [reply packages/gatsby/cache-dir/loader.js]: It seems left here by mistake — the other comments with "Tell plugins that..." all are before an invocation of a browser API.

**Comment by KyleAMathews** [comment]
I think we've cracked the "Missing Resource for '/'" problem! Looking at sentry, this error has basically dropped off and the few that are still happening came from older releases (so people on service workers) https://sentry.io/organizations/gatsby-inc-9z/issues/1054226255/tags/release/

This seems ready to go.

**Comment by iamskar** [comment]
Great stuff !!  Do we have a timeline on when this will be merged to master?

**Comment by wardpeet** [comment]
Writing tests and than we can merge. If it's not today than it will be tomorrow. I'll try my best to finish it today

**Comment by m-allanson** [comment]
🎉 

@wardpeet looks like there are some merge conflicts to fix up?

**Comment by KyleAMathews** [comment]
```
Successfully published:
 - gatsby-cli@2.7.4
 - gatsby@2.11.1
```

**Comment by AlexFoster009** [comment]
Im getting a 403 hosting my site with S3 / cloudfront. Its adding an extra / before page-data.json causing issues. Works as expected on localhost:8000 but on my cloudfront distribution its adding a slash.

local: http://localhost:8000/page-data/apply-now/page-data.json
cloudfront: https://d75attx91ridx.cloudfront.net/page-data//page-data.json

as you can see there is a missing bit there. Anyone have a solution? im also not sure if this is the correct place to post this, if it isn't could someone point me into the right direction?

**Comment by wardpeet** [comment]
Seems like your login page uses gatsby-link where it should be just a regular a tag. Gatsby-link only works for gatsby routes.

image

**Pull request merged**